py/server/deephaven/experimental/data_index.py

```diff
@@ -0,0 +1,90 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+"""This module provides the ability to create, check, and retrieve DataIndex objects from Deephaven tables."""
+from typing import List, Optional
+
+import jpy
+
+from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.jcompat import j_list_to_list
+from deephaven.table import Table
+
+_JDataIndexer = jpy.get_type("io.deephaven.engine.table.impl.indexer.DataIndexer")
+_JDataIndex = jpy.get_type("io.deephaven.engine.table.DataIndex")
+
+class DataIndex(JObjectWrapper):
+    """A DataIndex represents a Deephaven data index object which is built on one or more indexed (key) column(s) of
+    A Deephaven table."""
+    j_object_type = _JDataIndex
+
+    def __init__(self, j_data_index: jpy.JType):
+        self._j_data_indexer = j_data_index
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self._j_data_indexer
+
+    @property
+    def keys(self) -> List[str]:
+        """the key column names of the DataIndex.
+
+        Returns:
+            the key columns
+        """
+        return j_list_to_list(self._j_data_indexer.keyColumnNames())
+
+    @property
+    def table(self) -> Table:
+        """the backing table of the DataIndex.
+
+        Returns:
+            the backing table
+        """
+        return Table(self._j_data_indexer.table())
+
+
+
+def has_data_index(table: Table, key_cols: List[str]) -> bool:
+    """Checks if a table currently has a DataIndex for the given key columns.
+
+    Args:
+        table (Table): the table to check
+        key_cols (List[str]): the key columns to check
+
+    Returns:
+        bool: True if the table has a DataIndex, False otherwise
+    """
+    return _JDataIndexer.hasDataIndex(table.j_table, key_cols)
+
+def get_data_index(table: Table, key_cols: List[str]) -> Optional[DataIndex]:
+    """Gets a DataIndex for the given key columns. Returns None if the DataIndex does not exist.
+
+    Args:
+        table (Table): the table to get the DataIndex from
+        key_cols (List[str]): the key columns
+
+    Returns:
+        DataIndex: the DataIndex
+    """
+    j_di = _JDataIndexer.getDataIndex(table.j_table, key_cols)
+    return DataIndex(j_di) if j_di else None
+
+def create_data_index(table: Table, key_cols: List[str]) -> DataIndex:
```

Comment:
I quibble with the naming, not because it's incorrect but that it differs so much from the Java method.  The accompanying documentation does make it clear that this will return an existing index so it's probably OK.

Reply:
I don't like the name `get_or_create_data_index` as it is too verbose to be really Pythonic. But I'll leave it to @chipkent to have the final say.

Reply:
I agree that there is potentially some ambiguity that `create_` might not actually create anything.  Having said that, I don't know if I'm opposed to the name.  Let me offer a suggestion that is farther from the Java but possibly more pythonic and less ambiguous for the user.  `index_table`.

Reply:
Now that I'm looking at the other functions, I see that there is a `get_data_index` method.  Maybe it makes sense to keep this one as `create_data_index`.  I'm concerned that `get_or_create_data_index` may deviate from being pythonic.

Another pythonic alternative could be something like:
```python
def data_index(table: Table, key_cols: List[str], create: bool = False) -> Optional[DataIndex]:
```

Reply:
Is it more "pythonic" to imply via the function name that we're doing something we may not be doing?

Reply:
`ensure_data_index`?

Reply:
To be fair, unless users are simply poking at a table to see what it has, `get_or_create_data_index` is the only method they need or should use.
`has_data_index` is almost useless, because it's a snapshot of current state, and there's no guarantee that it remains true.
`get_data_index` is what most engine operations need to use.

Maybe we only really need to wrap `get_or_create_data_index`?

Reply:
I am OK with renaming it to `ensure_data_index` and I don't have a strong opinion on whether to keep `has_data_index` and `get_data_index`.  @chipkent @lbooker42 Your thoughts?

Reply:
At the Python sync meeting, we have agreed on changing it to:
```python
def data_index(table: Table, key_cols: List[str], create_if_absent: bool = True) ->DataIndex:
```